File: WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderSimplePostContainerView.java
package org.wordpress.android.ui.reader.views;

import android.content.Context;
import android.util.AttributeSet;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.wordpress.android.R;
import org.wordpress.android.ui.reader.ReaderInterfaces.OnFollowListener;
import org.wordpress.android.ui.reader.models.ReaderSimplePost;
import org.wordpress.android.ui.reader.models.ReaderSimplePostList;
import org.wordpress.android.util.analytics.AnalyticsUtils;

/**
 * used by the detail view to display related posts, which can be either local (related posts
 * from the same site as the source post) or global (related posts from across wp.com)
 */
public class ReaderSimplePostContainerView extends LinearLayout {
    private OnFollowListener mFollowListener;

    private final ReaderSimplePostList mSimplePostList = new ReaderSimplePostList();

    public ReaderSimplePostContainerView(Context context) {
        super(context);
        initView(context);
    }

    public ReaderSimplePostContainerView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initView(context);
    }

    public ReaderSimplePostContainerView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView(context);
    }

    public ReaderSimplePostContainerView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        initView(context);
    }

    private void initView(Context context) {
        inflate(context, R.layout.reader_simple_posts_container_view, this);
    }

    public void showPosts(ReaderSimplePostList posts,
                          String siteName,
                          boolean isGlobal,
                          ReaderSimplePostView.OnSimplePostClickListener listener) {
        mSimplePostList.clear();
        mSimplePostList.addAll(posts);

        ViewGroup container = (ViewGroup) findViewById(R.id.container_related_posts);
        container.removeAllViews();

        // nothing more to do if passed list is empty
        if (mSimplePostList.size() == 0) {
            return;
        }

        // add a view for each post
        for (int index = 0; index < mSimplePostList.size(); index++) {
            ReaderSimplePost relatedPost = mSimplePostList.get(index);
            ReaderSimplePostView postView = new ReaderSimplePostView(getContext());
            postView.setOnFollowListener(mFollowListener);
            postView.showPost(relatedPost, container, isGlobal, listener);
        }

        // make sure the label for these posts has the correct caption
        TextView label = findViewById(R.id.text_related_posts_label);
        if (isGlobal) {
            label.setText(getContext().getString(R.string.reader_label_global_related_posts));
        } else {
            label.setText(String.format(getContext().getString(R.string.reader_label_local_related_posts), siteName));
        }
    }

    public void setOnFollowListener(OnFollowListener listener) {
        mFollowListener = listener;
    }

    /*
     * called by reader detail when scrolled into view, tracks railcar events for each post
     */
    public void trackRailcarRender() {
        for (ReaderSimplePost post : mSimplePostList) {
            AnalyticsUtils.trackRailcarRender(post.getRailcarJson());
        }
    }
}

File: WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderSimplePostContainerView.kt
package org.wordpress.android.ui.reader.views

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.LinearLayout
import androidx.recyclerview.widget.LinearLayoutManager
import org.wordpress.android.WordPress
import org.wordpress.android.databinding.ReaderSimplePostsContainerViewBinding
import org.wordpress.android.ui.reader.adapters.ReaderRelatedPostsAdapter
import org.wordpress.android.ui.reader.tracker.ReaderTracker
import org.wordpress.android.ui.reader.viewmodels.ReaderPostDetailViewModel.UiState.ReaderPostDetailsUiState.RelatedPostsUiState
import org.wordpress.android.ui.utils.UiHelpers
import org.wordpress.android.util.image.ImageManager
import javax.inject.Inject

/**
 * used by the detail view to display related posts, which can be either local (related posts
 * from the same site as the source post) or global (related posts from across wp.com)
 */
class ReaderSimplePostContainerView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {
    private val binding = ReaderSimplePostsContainerViewBinding.inflate(LayoutInflater.from(context), this, true)
    private val railcarJsonStrings = mutableListOf<String?>()

    @Inject lateinit var uiHelpers: UiHelpers
    @Inject lateinit var imageManager: ImageManager
    @Inject lateinit var readerTracker: ReaderTracker

    init {
        (context.applicationContext as WordPress).component().inject(this)
        binding.initRecyclerView(context)
    }

    private fun ReaderSimplePostsContainerViewBinding.initRecyclerView(context: Context) {
        recyclerView.layoutManager = LinearLayoutManager(context)
        recyclerView.adapter = ReaderRelatedPostsAdapter(uiHelpers, imageManager)
    }

    fun showPosts(state: RelatedPostsUiState) = with(binding) {
        if (state.cards?.size == 0) return

        railcarJsonStrings.clear()
        railcarJsonStrings.addAll(state.railcarJsonStrings)

        state.cards?.let { (recyclerView.adapter as ReaderRelatedPostsAdapter).update(it) }
        uiHelpers.setTextOrHide(textRelatedPostsLabel, state.headerLabel)
    }

    /*
     * called by reader detail when scrolled into view, tracks railcar events for each post
     */
    @Suppress("ForbiddenComment")
    fun trackRailcarRender() { // TODO: move tracking to view model
        for (railcarJson in railcarJsonStrings) {
            railcarJson?.let { readerTracker.trackRailcar(it) }
        }
    }
}
